File: string_algorithms/week1_suffix_trees/4_suffix_tree/suffix_tree.cpp
#include <deque>
#include <iostream>
#include <map>
#include <optional>
#include <string>
#include <vector>

using namespace std;

namespace suffix_tree {
struct Node {
  optional<string> label;
  map<char, Node> outgoing_edges;

  Node(const string& str = "") : label(str) {}
};

void AddSuffix(Node& root, const string& suffix) {
  Node* currentNode = &root;

  for (char symbol : suffix) {
    if (currentNode->outgoing_edges.find(symbol) == currentNode->outgoing_edges.end())
      currentNode->outgoing_edges[symbol] = Node();

    currentNode = &currentNode->outgoing_edges[symbol];
  }
}

void TraverseTree(const Node& root, vector<string>& result) {
  deque<const Node*> dq;
  dq.push_back(&root);

  while (!dq.empty()) {
    const Node* u = dq.front();
    dq.pop_front();

    if (u->label.has_value())
      result.push_back(u->label.value());

    for (const auto& pair : u->outgoing_edges)
      dq.push_back(&(pair.second));
  }
}

void ComputeSuffixTreeEdges(const string& s, vector<string>& result) {
  Node root;
  root.label = nullopt;
  root.outgoing_edges[s[0]] = Node();
  root.outgoing_edges[s[0]].label = s;

  for (int i = 1; i < s.length(); ++i) {
    Node* cur = &root;
    int j = i;
    while (j < s.length()) {
      if (cur->outgoing_edges.find(s[j]) != cur->outgoing_edges.end()) {
        Node& child = cur->outgoing_edges[s[j]];
        string label = child.label.value_or("");
        int k = j + 1;
        while (k - j < label.length() && s[k] == label[k - j])
          k += 1;

        if (k - j == label.length()) {
          cur = &child;
          j = k;
        } else {
          char cExist = label[k - j];
          char cNew = s[k];
          Node mid;
          mid.label = label.substr(0, k - j);
          mid.outgoing_edges[cNew] = Node(s.substr(k));
          child.label = label.substr(k - j);
          mid.outgoing_edges[cExist] = child;
          cur->outgoing_edges[s[j]] = mid;
        }
      } else
        cur->outgoing_edges[s[j]] = Node(s.substr(j));
    }
  }

  TraverseTree(root, result);
}
}  // namespace suffix_tree

int main() {
  string text;
  cin >> text;

  vector<string> edges;
  suffix_tree::ComputeSuffixTreeEdges(text, edges);

  for (const string& edge : edges)
    cout << edge << endl;

  return 0;
}

File: string_algorithms/week1_suffix_trees/4_suffix_tree/suffix_tree.py
from collections import deque
from typing import Dict, Optional


class Node:
    def __init__(self, label: Optional[str]) -> None:
        self.label = label
        self.outgoing_edges: Dict[str, Node] = {}


class SuffixTree:
    def __init__(self, s: str) -> None:  # noqa: C901
        self.root = Node(None)
        self.root.outgoing_edges[s[0]] = Node(s)

        for i in range(1, len(s)):
            cur = self.root
            j = i
            while j < len(s):
                if s[j] in cur.outgoing_edges:
                    child = cur.outgoing_edges[s[j]]
                    label = child.label or ""
                    k = j + 1
                    while k - j < len(label) and s[k] == label[k - j]:
                        k += 1
                    if k - j == len(label):
                        cur = child
                        j = k
                    else:
                        existing, new = label[k - j], s[k]
                        mid = Node(label[: k - j])
                        mid.outgoing_edges[new] = Node(s[k:])
                        child.label = label[k - j :]
                        mid.outgoing_edges[existing] = child
                        cur.outgoing_edges[s[j]] = mid
                else:
                    cur.outgoing_edges[s[j]] = Node(s[j:])

    def __str__(self) -> str:
        res = []
        dq: deque = deque()
        dq.append(self.root)
        while dq:
            u = dq.popleft()
            if u != self.root:
                res.append(u.label)
            dq.extend(u.outgoing_edges.values())
        return "\n".join(res)


if __name__ == "__main__":
    text = input()
    stree = SuffixTree(text)
    print(stree)
